extensions/big-data-cluster/src/bigDataCluster/dialog/addControllerDialog.ts

```diff
@@ -0,0 +1,501 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+import * as azdata from 'azdata';
+import * as nls from 'vscode-nls';
+import { ClusterController, IEndPoint, IControllerError } from '../controller/clusterController';
+import { ControllerTreeDataProvider } from '../tree/controllerTreeDataProvider';
+import { TreeNode } from '../tree/treeNode';
+
+const localize = nls.loadMessageBundle();
+
+export class AddControllerDialogModel {
+	constructor(
+		public treeDataProvider: ControllerTreeDataProvider,
+		public node?: TreeNode,
+		public prefilledUrl?: string,
+		public prefilledUsername?: string,
+		public prefilledPassword?: string,
+		public prefilledRememberPassword?: boolean
+	) {
+		this.prefilledUrl = prefilledUrl || (node && node['url']);
+		this.prefilledUsername = prefilledUsername || (node && node['username']);
+		this.prefilledPassword = prefilledPassword || (node && node['password']);
+		this.prefilledRememberPassword = prefilledRememberPassword || (node && node['rememberPassword']);
+	}
+
+	public async onComplete(url: string, username: string, password: string, rememberPassword: boolean): Promise<void> {
+		let clusterController = new ClusterController();
+		let response = await clusterController.getEndPoints(url, username, password, true);
+		if (response && response.request) {
+			let masterInstance: IEndPoint = undefined;
+			if (response.endPoints) {
+				masterInstance = response.endPoints.find(e => e.name && e.name === 'sql-server-master');
+			}
+			this.treeDataProvider.addController(response.request.url, response.request.username,
+				response.request.password, rememberPassword, masterInstance);
+			await this.treeDataProvider.saveControllers();
+		}
+	}
+
+	public async onError(error: IControllerError): Promise<void> {
+		// implement
+	}
+
+	public async onCancel(): Promise<void> {
+		if (this.node) {
+			this.node.refresh();
+		}
+	}
+}
+
+export class AddControllerDialog {
+
+	private readonly DialogTitle: string = localize('textAddNewController', 'Add New Controller');
```

Comment:
>private readonly DialogTitle [](start = 1, length = 28)

These can just be constants.

Reply:
Or just be declared inline to reduce the lines of code in the file. That's the standard pattern in our repo, unless you need to use the same string twice.

Reply:
removed.